test/broadway_test.exs

```diff
@@ -1760,6 +1764,157 @@ defmodule BroadwayTest do
+           %Message{data: 4, acknowledger: {CallerAcknowledger, {self(), :ref}, :unused}},
+           %Message{data: 5, acknowledger: {CallerAcknowledger, {self(), :ref}, :unused}}
+         ]}
+      )
+
+      assert_receive {:handle_message_called, %Message{data: 1}, timestamp1}
+      assert_receive {:handle_message_called, %Message{data: 2}, _timestamp2}
+      assert_receive {:handle_message_called, %Message{data: 3}, timestamp3}
+      assert_receive {:handle_message_called, %Message{data: 4}, _timestamp4}
+      assert_receive {:handle_message_called, %Message{data: 5}, timestamp5}
+
+      assert_receive {:ack, :ref, [_, _], []}
+      assert_receive {:ack, :ref, [_, _], []}
+      assert_receive {:ack, :ref, [_], []}
+
+      # To be safe, we're saying that these messages have to be at least 10ms apart
+      # even if the interval is 50ms. This is because the time when the processor
+      # receives the message is not the time when the producer produced it.
+      assert System.convert_time_unit(timestamp3 - timestamp1, :native, :millisecond) >= 10
+      assert System.convert_time_unit(timestamp5 - timestamp3, :native, :millisecond) >= 10
+    end
+
+    test "buffers demand and flushes it when the rate limiting is lifted" do
+      broadway_name = new_unique_name()
+      test_pid = self()
+
+      handle_message = fn message, _ ->
+        send(test_pid, {:handle_message_called, message})
+        message
+      end
+
+      {:ok, _broadway} =
+        Broadway.start_link(CustomHandlers,
+          name: broadway_name,
+          producer: [
+            module: {ManualProducer, %{test_pid: test_pid}},
+            stages: 1,
+            # We use a long rate limiting interval so that we can trigger the rate limiting
+            # manually from this test.
+            rate_limiting: [allowed_messages: 2, interval: 10000]
+          ],
+          processors: [default: [stages: 1, max_demand: 2]],
+          context: %{handle_message: handle_message}
+        )
+
+      send(
+        get_producer(broadway_name),
+        {:push_messages,
+         [
+           %Message{data: 1, acknowledger: {CallerAcknowledger, {self(), :ref}, :unused}},
+           %Message{data: 2, acknowledger: {CallerAcknowledger, {self(), :ref}, :unused}},
+           %Message{data: 3, acknowledger: {CallerAcknowledger, {self(), :ref}, :unused}}
+         ]}
+      )
+
+      # First, we assert that handle_demand is called and that two events are emitted.
+      assert_receive {:handle_demand_called, _demand}
+      assert_receive {:handle_message_called, %Message{data: 1}}
+      assert_receive {:handle_message_called, %Message{data: 2}}
+
+      # Now, since the rate limiting interval is long, we can assert that
+      # handle_demand is not called and that the third message is not emitted yet.
+      refute_receive {:handle_demand_called, _demand}
+      refute_receive {:handle_message_called, %Message{data: 3}}, 500
```

Comment:
`refute_receive` is very expensive. I recommend replacing all of them in these new tests by `refute_received`. It means there is a chance the message may arrive later but it will still catch invalid tests from time to time. Something else that you could do is to include the timestamp on handle_demand_called and handle_message_called and before you send  the `:reset_limit`message, you read the timestamp and ensure the messages were delivered later. This provides stronger guarantees than refute_receive without the waiting.